File: android/src/main/java/com/eliazavatta/plugins/unityads/Unityads.java
package com.eliazavatta.plugins.unityads;

import android.content.Context;
import android.util.Log;
import com.unity3d.ads.IUnityAdsInitializationListener;
import com.unity3d.ads.IUnityAdsLoadListener;
import com.unity3d.ads.IUnityAdsShowListener;
import com.unity3d.ads.UnityAds;
import com.unity3d.ads.UnityAdsShowOptions;

public class Unityads {

    private static final String TAG = "UnityAds";

    private boolean isInitialized = false;
    private boolean testMode = false;
    private String currentRewardedPlacementId;
    private String currentInterstitialPlacementId;
    private boolean rewardedVideoLoaded = false;
    private boolean interstitialLoaded = false;

    // Callback interfaces
    public interface InitializationCallback {
        void onSuccess();
        void onError(String error);
    }

    public interface AdLoadCallback {
        void onAdLoaded();
        void onError(String error);
    }

    public interface RewardedVideoCallback {
        void onAdShown();
        void onRewardEarned(String rewardType, int rewardAmount);
        void onAdClosed();
        void onError(String error);
    }

    public interface InterstitialCallback {
        void onAdShown();
        void onAdClosed();
        void onError(String error);
    }

    public void initialize(Context context, String gameId, boolean testMode, InitializationCallback callback) {
        try {
            Log.d(TAG, "Initializing Unity Ads with Game ID: " + gameId);

            this.testMode = testMode;

            IUnityAdsInitializationListener initializationListener = new IUnityAdsInitializationListener() {
                @Override
                public void onInitializationComplete() {
                    Log.d(TAG, "Unity Ads initialized successfully");
                    isInitialized = true;
                    callback.onSuccess();
                }

                @Override
                public void onInitializationFailed(UnityAds.UnityAdsInitializationError error, String message) {
                    Log.e(TAG, "Unity Ads initialization failed: " + message);
                    callback.onError("Initialization failed: " + message);
                }
            };

            // Initialize Unity Ads
            UnityAds.initialize(context, gameId, testMode, initializationListener);
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize Unity Ads", e);
            callback.onError("Initialization failed: " + e.getMessage());
        }
    }

    public void loadRewardedVideo(String placementId, AdLoadCallback callback) {
        if (!isInitialized) {
            callback.onError("Unity Ads not initialized");
            return;
        }

        try {
            Log.d(TAG, "Loading rewarded video with placement ID: " + placementId);

            currentRewardedPlacementId = placementId;
            rewardedVideoLoaded = false;

            IUnityAdsLoadListener loadListener = new IUnityAdsLoadListener() {
                @Override
                public void onUnityAdsAdLoaded(String placementId) {
                    Log.d(TAG, "Rewarded video loaded successfully");
                    rewardedVideoLoaded = true;
                    callback.onAdLoaded();
                }

                @Override
                public void onUnityAdsFailedToLoad(String placementId, UnityAds.UnityAdsLoadError error, String message) {
                    Log.e(TAG, "Rewarded video failed to load: " + message);
                    rewardedVideoLoaded = false;
                    callback.onError("Failed to load rewarded video: " + message);
                }
            };

            UnityAds.load(placementId, loadListener);
        } catch (Exception e) {
            Log.e(TAG, "Error loading rewarded video", e);
            callback.onError("Error loading rewarded video: " + e.getMessage());
        }
    }

    public void showRewardedVideo(android.app.Activity activity, RewardedVideoCallback callback) {
        if (!isInitialized) {
            callback.onError("Unity Ads not initialized");
            return;
        }

        if (currentRewardedPlacementId == null || !rewardedVideoLoaded) {
            callback.onError("Rewarded video not loaded");
            return;
        }

        try {
            Log.d(TAG, "Showing rewarded video");

            IUnityAdsShowListener showListener = new IUnityAdsShowListener() {
                @Override
                public void onUnityAdsShowStart(String placementId) {
                    Log.d(TAG, "Rewarded video show started");
                    callback.onAdShown();
                }

                @Override
                public void onUnityAdsShowClick(String placementId) {
                    Log.d(TAG, "Rewarded video clicked");
                }

                @Override
                public void onUnityAdsShowComplete(String placementId, UnityAds.UnityAdsShowCompletionState state) {
                    Log.d(TAG, "Rewarded video show completed with state: " + state);

                    if (state == UnityAds.UnityAdsShowCompletionState.COMPLETED) {
                        // User watched the full video, give reward
                        callback.onRewardEarned("coins", 1);
                    } else {
                        // User skipped or ad was not completed
                        callback.onAdClosed();
                    }

                    rewardedVideoLoaded = false; // Reset loaded state
                }

                @Override
                public void onUnityAdsShowFailure(String placementId, UnityAds.UnityAdsShowError error, String message) {
                    Log.e(TAG, "Rewarded video show failed: " + message);
                    callback.onError("Failed to show rewarded video: " + message);
                    rewardedVideoLoaded = false;
                }
            };

            UnityAds.show(activity, currentRewardedPlacementId, new UnityAdsShowOptions(), showListener);
        } catch (Exception e) {
            Log.e(TAG, "Error showing rewarded video", e);
            callback.onError("Error showing rewarded video: " + e.getMessage());
        }
    }

    public boolean isRewardedVideoLoaded() {
        return rewardedVideoLoaded && currentRewardedPlacementId != null;
    }

    public void loadInterstitial(String placementId, AdLoadCallback callback) {
        if (!isInitialized) {
            callback.onError("Unity Ads not initialized");
            return;
        }

        try {
            Log.d(TAG, "Loading interstitial with placement ID: " + placementId);

            currentInterstitialPlacementId = placementId;
            interstitialLoaded = false;

            IUnityAdsLoadListener loadListener = new IUnityAdsLoadListener() {
                @Override
                public void onUnityAdsAdLoaded(String placementId) {
                    Log.d(TAG, "Interstitial loaded successfully");
                    interstitialLoaded = true;
                    callback.onAdLoaded();
                }

                @Override
                public void onUnityAdsFailedToLoad(String placementId, UnityAds.UnityAdsLoadError error, String message) {
                    Log.e(TAG, "Interstitial failed to load: " + message);
                    interstitialLoaded = false;
                    callback.onError("Failed to load interstitial: " + message);
                }
            };

            UnityAds.load(placementId, loadListener);
        } catch (Exception e) {
            Log.e(TAG, "Error loading interstitial", e);
            callback.onError("Error loading interstitial: " + e.getMessage());
        }
    }

    public void showInterstitial(android.app.Activity activity, InterstitialCallback callback) {
        if (!isInitialized) {
            callback.onError("Unity Ads not initialized");
            return;
        }

        if (currentInterstitialPlacementId == null || !interstitialLoaded) {
            callback.onError("Interstitial not loaded");
            return;
        }

        try {
            Log.d(TAG, "Showing interstitial");

            IUnityAdsShowListener showListener = new IUnityAdsShowListener() {
                @Override
                public void onUnityAdsShowStart(String placementId) {
                    Log.d(TAG, "Interstitial show started");
                    callback.onAdShown();
                }

                @Override
                public void onUnityAdsShowClick(String placementId) {
                    Log.d(TAG, "Interstitial clicked");
                }

                @Override
                public void onUnityAdsShowComplete(String placementId, UnityAds.UnityAdsShowCompletionState state) {
                    Log.d(TAG, "Interstitial show completed");
                    callback.onAdClosed();
                    interstitialLoaded = false; // Reset loaded state
                }

                @Override
                public void onUnityAdsShowFailure(String placementId, UnityAds.UnityAdsShowError error, String message) {
                    Log.e(TAG, "Interstitial show failed: " + message);
                    callback.onError("Failed to show interstitial: " + message);
                    interstitialLoaded = false;
                }
            };

            UnityAds.show(activity, currentInterstitialPlacementId, new UnityAdsShowOptions(), showListener);
        } catch (Exception e) {
            Log.e(TAG, "Error showing interstitial", e);
            callback.onError("Error showing interstitial: " + e.getMessage());
        }
    }

    public boolean isInterstitialLoaded() {
        return interstitialLoaded && currentInterstitialPlacementId != null;
    }

    public void setTestMode(boolean enabled) {
        Log.d(TAG, "Setting test mode: " + enabled);
        this.testMode = enabled;
        // Unity Ads test mode is set during initialization
    }

    public String getVersion() {
        try {
            return UnityAds.getVersion();
        } catch (Exception e) {
            Log.e(TAG, "Error getting Unity Ads version", e);
            return "unknown";
        }
    }
}

File: ios/Sources/UnityadsPlugin/Unityads.swift
import Foundation
import UnityAds

@objc public class Unityads: NSObject {
    private var isInitialized = false
    private var testMode = false
    private var currentRewardedPlacementId: String?
    private var currentInterstitialPlacementId: String?
    private var rewardedVideoLoaded = false
    private var interstitialLoaded = false
    
    // Callback types
    typealias InitializationCallback = (Bool, String?) -> Void
    typealias AdLoadCallback = (Bool, String?) -> Void
    typealias RewardedVideoCallback = (Bool, [String: Any]?, String?) -> Void
    typealias InterstitialCallback = (Bool, String?) -> Void
    
    func initialize(gameId: String, testMode: Bool, callback: @escaping InitializationCallback) {
        print("[UnityAds] Initializing with Game ID: \(gameId)")
        
        self.testMode = testMode
        
        UnityAds.initialize(gameId, testMode: testMode) { [weak self] in
            print("[UnityAds] Initialized successfully")
            self?.isInitialized = true
            callback(true, nil)
        } errorHandler: { [weak self] error in
            print("[UnityAds] Initialization failed: \(error.localizedDescription)")
            callback(false, error.localizedDescription)
        }
    }
    
    func loadRewardedVideo(placementId: String, callback: @escaping AdLoadCallback) {
        guard isInitialized else {
            callback(false, "Unity Ads not initialized")
            return
        }
        
        print("[UnityAds] Loading rewarded video with placement ID: \(placementId)")
        
        currentRewardedPlacementId = placementId
        rewardedVideoLoaded = false
        
        UnityAds.load(placementId, loadDelegate: RewardedVideoLoadDelegate(callback: callback, parent: self))
    }
    
    func showRewardedVideo(callback: @escaping RewardedVideoCallback) {
        guard isInitialized else {
            callback(false, nil, "Unity Ads not initialized")
            return
        }
        
        guard let placementId = currentRewardedPlacementId, rewardedVideoLoaded else {
            callback(false, nil, "Rewarded video not loaded")
            return
        }
        
        print("[UnityAds] Showing rewarded video")
        
        UnityAds.show(UIApplication.shared.windows.first?.rootViewController, placementId: placementId, showDelegate: RewardedVideoShowDelegate(callback: callback, parent: self))
    }
    
    func isRewardedVideoLoaded() -> Bool {
        return rewardedVideoLoaded && currentRewardedPlacementId != nil
    }
    
    func loadInterstitial(placementId: String, callback: @escaping AdLoadCallback) {
        guard isInitialized else {
            callback(false, "Unity Ads not initialized")
            return
        }
        
        print("[UnityAds] Loading interstitial with placement ID: \(placementId)")
        
        currentInterstitialPlacementId = placementId
        interstitialLoaded = false
        
        UnityAds.load(placementId, loadDelegate: InterstitialLoadDelegate(callback: callback, parent: self))
    }
    
    func showInterstitial(callback: @escaping InterstitialCallback) {
        guard isInitialized else {
            callback(false, "Unity Ads not initialized")
            return
        }
        
        guard let placementId = currentInterstitialPlacementId, interstitialLoaded else {
            callback(false, "Interstitial not loaded")
            return
        }
        
        print("[UnityAds] Showing interstitial")
        
        UnityAds.show(UIApplication.shared.windows.first?.rootViewController, placementId: placementId, showDelegate: InterstitialShowDelegate(callback: callback, parent: self))
    }
    
    func isInterstitialLoaded() -> Bool {
        return interstitialLoaded && currentInterstitialPlacementId != nil
    }
    
    func setTestMode(enabled: Bool) {
        print("[UnityAds] Setting test mode: \(enabled)")
        self.testMode = enabled
        // Unity Ads test mode is set during initialization
    }
    
    func getVersion() -> String {
        return UnityAds.getVersion() ?? "unknown"
    }
}

// MARK: - Load Delegates

class RewardedVideoLoadDelegate: NSObject, UnityAdsLoadDelegate {
    private let callback: Unityads.AdLoadCallback
    private weak var parent: Unityads?
    
    init(callback: @escaping Unityads.AdLoadCallback, parent: Unityads) {
        self.callback = callback
        self.parent = parent
    }
    
    func unityAdsAdLoaded(_ placementId: String) {
        print("[UnityAds] Rewarded video loaded successfully")
        parent?.rewardedVideoLoaded = true
        callback(true, nil)
    }
    
    func unityAdsAdFailedToLoad(_ placementId: String, withError error: UnityAdsLoadError, withMessage message: String) {
        print("[UnityAds] Rewarded video failed to load: \(message)")
        parent?.rewardedVideoLoaded = false
        callback(false, "Failed to load rewarded video: \(message)")
    }
}

class InterstitialLoadDelegate: NSObject, UnityAdsLoadDelegate {
    private let callback: Unityads.AdLoadCallback
    private weak var parent: Unityads?
    
    init(callback: @escaping Unityads.AdLoadCallback, parent: Unityads) {
        self.callback = callback
        self.parent = parent
    }
    
    func unityAdsAdLoaded(_ placementId: String) {
        print("[UnityAds] Interstitial loaded successfully")
        parent?.interstitialLoaded = true
        callback(true, nil)
    }
    
    func unityAdsAdFailedToLoad(_ placementId: String, withError error: UnityAdsLoadError, withMessage message: String) {
        print("[UnityAds] Interstitial failed to load: \(message)")
        parent?.interstitialLoaded = false
        callback(false, "Failed to load interstitial: \(message)")
    }
}

// MARK: - Show Delegates

class RewardedVideoShowDelegate: NSObject, UnityAdsShowDelegate {
    private let callback: Unityads.RewardedVideoCallback
    private weak var parent: Unityads?
    
    init(callback: @escaping Unityads.RewardedVideoCallback, parent: Unityads) {
        self.callback = callback
        self.parent = parent
    }
    
    func unityAdsShowStart(_ placementId: String) {
        print("[UnityAds] Rewarded video show started")
    }
    
    func unityAdsShowClick(_ placementId: String) {
        print("[UnityAds] Rewarded video clicked")
    }
    
    func unityAdsShowComplete(_ placementId: String, withFinishState state: UnityAdsShowCompletionState) {
        print("[UnityAds] Rewarded video show completed with state: \(state.rawValue)")
        
        parent?.rewardedVideoLoaded = false // Reset loaded state
        
        if state == .completed {
            // User watched the full video, give reward
            let reward = ["type": "coins", "amount": 1] as [String : Any]
            callback(true, reward, nil)
        } else {
            // User skipped or ad was not completed
            callback(false, nil, nil)
        }
    }
    
    func unityAdsShowFailed(_ placementId: String, withError error: UnityAdsShowError, withMessage message: String) {
        print("[UnityAds] Rewarded video show failed: \(message)")
        parent?.rewardedVideoLoaded = false
        callback(false, nil, "Failed to show rewarded video: \(message)")
    }
}

class InterstitialShowDelegate: NSObject, UnityAdsShowDelegate {
    private let callback: Unityads.InterstitialCallback
    private weak var parent: Unityads?
    
    init(callback: @escaping Unityads.InterstitialCallback, parent: Unityads) {
        self.callback = callback
        self.parent = parent
    }
    
    func unityAdsShowStart(_ placementId: String) {
        print("[UnityAds] Interstitial show started")
        callback(true, nil)
    }
    
    func unityAdsShowClick(_ placementId: String) {
        print("[UnityAds] Interstitial clicked")
    }
    
    func unityAdsShowComplete(_ placementId: String, withFinishState state: UnityAdsShowCompletionState) {
        print("[UnityAds] Interstitial show completed")
        parent?.interstitialLoaded = false // Reset loaded state
    }
    
    func unityAdsShowFailed(_ placementId: String, withError error: UnityAdsShowError, withMessage message: String) {
        print("[UnityAds] Interstitial show failed: \(message)")
        parent?.interstitialLoaded = false
        callback(false, "Failed to show interstitial: \(message)")
    }
}
